pageserver/src/restore_local_repo.rs

```diff
@@ -0,0 +1,460 @@
+        if !PathBuf::from(&path).exists() {
+            path = path + ".partial";
+        }
+
+        // Slurp the WAL file
+        let open_result = File::open(&path);
+        if let Err(e) = open_result {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                break;
+            }
+            return Err(e)?;
+        }
+        let mut file = open_result.unwrap();
+
+        if offset > 0 {
+            file.seek(SeekFrom::Start(offset as u64))?;
+        }
+
+        let mut buf = Vec::new();
+        let nread = file.read_to_end(&mut buf)?;
+        if nread != 16 * 1024 * 1024 - offset as usize {
+            // Maybe allow this for .partial files?
+            error!("read only {} bytes from WAL file", nread);
+        }
+        waldecoder.feed_bytes(&buf);
+
+        let mut nrecords = 0;
+        loop {
+            let rec = waldecoder.poll_decode();
+            if rec.is_err() {
+                // Assume that an error means we've reached the end of
+                // a partial WAL record. So that's ok.
+                break;
+            }
+            if let Some((lsn, recdata)) = rec.unwrap() {
+                let decoded =
+                    crate::waldecoder::decode_wal_record(recdata.clone());
+
+                // Put the WAL record to the page cache. We make a separate copy of
+                // it for every block it modifies. (The actual WAL record is kept in
+                // a Bytes, which uses a reference counter for the underlying buffer,
+                // so having multiple copies of it doesn't cost that much)
+                for blk in decoded.blocks.iter() {
+                    let tag = BufferTag {
+                        spcnode: blk.rnode_spcnode,
+                        dbnode: blk.rnode_dbnode,
+                        relnode: blk.rnode_relnode,
+                        forknum: blk.forknum as u8,
+                        blknum: blk.blkno,
+                    };
+
+                    let rec = page_cache::WALRecord {
+                        lsn: lsn,
+                        will_init: blk.will_init || blk.apply_image,
+                        rec: recdata.clone(),
+                    };
+
+                    pcache.put_wal_record(tag, rec);
+                }
+
+                // Now that this record has been handled, let the page cache know that
+                // it is up-to-date to this LSN
+                pcache.advance_last_valid_lsn(lsn);
+                last_lsn = lsn;
+            } else {
+                break;
+            }
+            nrecords += 1;
+        }
+
+        info!("restored {} records from WAL file {}", nrecords, filename);
+
+        segno += 1;
+        offset = 0;
+    }
+    info!("reached end of WAL at {:X}/{:X}", last_lsn >> 32, last_lsn & 0xffffffff);
+
+    Ok(())
+}
+
+// FIXME: copied from xlog_utils.rs
+pub const XLOG_FNAME_LEN: usize = 24;
+pub type XLogRecPtr = u64;
+pub type XLogSegNo = u64;
+pub type TimeLineID = u32;
```

Comment:
Should we make these newtypes so they can't be mixed up with one another?

E.g.
```
pub struct XlogRecPtr(pub u64);
```

Reply:
Yeah. XLogRecPtr is anther name for LSN, so https://github.com/zenithdb/zenith/issues/49 applies here. These are copy-pasted from `xlog_utils.rs`, so we should refactor these into some generic module. Perhaps just move xlog_utils.rs to a common crate. I'll push this as it is for now, but this certainly needs cleanup.

Reply:
I like the idea of a crate where we can drop useful functions/types/etc. that are likely to be shared.